src/repr/src/row/collection.rs

```diff
@@ -132,16 +142,48 @@ impl RowCollection {
     }
 
     /// "Sorts" the [`RowCollection`] by returning a sorted view over the collection.
-    pub fn sorted_view<F>(self, mut f: F) -> SortedRowCollection
+    pub fn sorted_view<F>(self, cmp: F) -> SortedRowCollection
     where
-        F: FnMut(&RowRef, &RowRef) -> std::cmp::Ordering,
+        F: Fn(&RowRef, &RowRef) -> std::cmp::Ordering,
     {
-        let mut view: Vec<_> = (0..self.metadata.len()).collect();
-        view.sort_by(|a, b| {
-            let (a_row, _) = self.get(*a).expect("index invalid?");
-            let (b_row, _) = self.get(*b).expect("index invalid?");
-            f(a_row, b_row)
-        });
+        let mut heap = BinaryHeap::new();
+
+        for index in 0..self.fingers.len() {
+            let start = (index == 0)
+                .then_some(0)
+                .unwrap_or_else(|| self.fingers[index - 1]);
+            let end = self.fingers[index];
+
+            // TODO: Remove
+            for j in start..end {
+                if j + 1 < end {
+                    assert_ne!(
+                        cmp(self.get(j).unwrap().0, self.get(j + 1).unwrap().0),
+                        std::cmp::Ordering::Greater
+                    );
+                }
+            }
+            if start < end {
+                heap.push(Reverse(Finger {
+                    collection: &self,
+                    cmp: &cmp,
+                    start,
+                    end,
+                }));
+            }
+        }
+
+        let mut view = Vec::with_capacity(self.metadata.len());
+
+        while let Some(Reverse(mut finger)) = heap.pop() {
+            view.push(finger.start);
+            finger.start += 1;
+            if finger.start < finger.end {
+                heap.push(Reverse(finger));
+            }
+        }
```

Comment:
It feels like there is a great opportunity to push this logic into `SortedRowCollection` or `SortedRowCollectionIter` maybe? i.e. as folks iterate through a row collection is when we do this streaming merge sort?

Reply:
Agreed! I assume we're going to iterate through the result rows only once when we send them over the wire, so we can avoid having the extra `view` buffer around and decoding the rows twice.

Reply:
Unfortunately, we need to iterate multiple times to determine the size of the whole result. I agree we shouldn't (and there is no deep reason we have to), but it requires more changes.